Self host port

arichiardi opened the pull request:
I have had some time to work on this and noticed that the changes are minimal actually.

You can try this works now in lumo with:

```
$ lumo -c `lein classpath`
cljs.user=> (require '[env-config.impl.report :as report])
nil
cljs.user=> (report/log-reports-if-needed! [[:debug "test"] [:warn "warn"]])
test
nil
warn
```

I have not tried the var reading part  and this is going to be my next step, while porting `cljs-oops`.

There is test failing here but it looks like it is just because the error message has different format. Let me know if is something that it is actually important to have right:

```
lein test :only env-config.tests.main/test-top-level-api

FAIL in (test-top-level-api) (main.clj:90)
make-config-with-logging
expected: (= [[:warn "env-config: unable to read-string from variable 'p/c' with value \"~#!@#xxx\", attempted to eval code: '#!@#xxx', got problem: EOF while reading."]] (clojure.core/deref reports-atom))
  actual: (not (= [[:warn "env-config: unable to read-string from variable 'p/c' with value \"~#!@#xxx\", attempted to eval code: '#!@#xxx', got problem: EOF while reading."]] [[:warn "env-config: unable to read-string from variable 'p/c' with value \"~#!@#xxx\", attempted to eval code: '#!@#xxx', got problem: No dispatch macro for: !."]]))

Ran 4 tests containing 15 assertions.
1 failures, 0 errors.
Tests failed.
```

I am looking forward for a review.

Thanks Darwin!

arichiardi committed:
Rename files to .cljc

arichiardi committed:
Adapt code for self-host compatibility

Not only this patch adds reader conditionals where necessary, but it
also introduces a custom read-string for code coercer and correct number
checks for both integer and double coercers.

darwin commented:
Looks good, but what about running the tests in self-hosted environment? Without this it will be prone to breaking if I touch it.

arichiardi commented:
Yep, [tach](https://github.com/mfikes/tach) will be necessary here.

arichiardi committed:
Instrument self-host tests

This patch uses mfikes/tach in order to add self-host tests and makes
the test main namespace ClojureScript compatible.

arichiardi commented:
Ok `tach` is in, everything fails :+1:

arichiardi commented:
I am really surprised about this failure:

```
FAIL in (test-top-level-api) (at evalmachine.<anonymous>:507:85)
make-config
expected: (= {:var 42} (make-config "project" {"project/var" "42"}))
  actual: (not (= {:var 42} {:var "42"}))
```

darwin commented:
I'm sorry. I won't have time to look at it until mid-next-week.

Let me know when you are done with this PR. I will then merge it and try to fix remaining issues.

arichiardi commented:
No problem @darwin! I am too going to be slow on this one :smile:

arichiardi commented:
Fixed some tests due to some weird JS/Clojure [behavior](https://github.com/binaryage/env-config/pull/2/commits/4e008f2015da58d9e916a344a6aa2d7d04bb9f4d#diff-e12ceebc500a3226e1d166507893b814R14) I did not know about.

Something is still failing but I have run out of time now, I'll get back to it.
For running, install `lumo` and then run `lein tach lumo`.

arichiardi commented on src/lib/env_config/impl/coercers.cljc:
```diff
@@ -17,32 +27,34 @@
-          (report/report-warning! (str "unable to read-string from " (make-var-description (meta path)) ", "
-                                       "attempted to eval code: '" code "', "
-                                       "got problem: " (.getMessage e) "."))
+        (->Coerced (edn/read-string code))
```
I switched to `edn/read-string` because of its safety and portability Of course the behavior changes here for the error message, so I will change the test as well.

arichiardi committed:
Convert one branch ifs to whens

arichiardi commented:
Fun! The port is complete! I wish there was an icon to put on the README for all self-host compatible repos :smile:

darwin commented:
@arichiardi I got rid of all reader conditionals by moving some code into platform-specific files and writing some helper macros.

Would be great if you could review my changes in master - if it worked for you, I would call it a day and cut the 0.2.0 release.

arichiardi commented:
@darwin checking.

arichiardi commented:
Ok @darwin no issue on my side, good job. If you want to specify that the lib is self-host compatible in the README it would be great, but I can probably PR it as well.

darwin commented:
Feel free to update the readme. I'm just going to mention it in release notes for 0.2.0.
